content/source/docs/extend/terraform-0.12-compatibility.html.md

```diff
@@ -0,0 +1,329 @@
+some small adjustments to provider code may be required to ensure the best
+possible compatibility with Terraform 0.12. The goal of this guide is to
+describe some common situations and how to address them.
+
+-> Terraform internally uses a separate versioning scheme for the provider
+protocol than for Terraform Core itself. The protocol version used in Terraform
+0.10 and 0.11 has version number 4, while Terraform 0.12 uses version number 5.
+You may see these version numbers in error messages in the event of a
+protocol compatibility problem.
+
+## Upgrading to the latest Terraform SDK
+
+At the time of the Terraform 0.12 release, the Terraform SDK is a set of
+sub-directories inside the Terraform Core repository. Therefore upgrading
+to the latest Terraform SDK involves upgrading all of the dependencies on
+Go packages with the prefix `github.com/hashicorp/terraform/` to a version
+with support for the new provider protocol.
+
+Terraform Core is now using
+[Go Modules](https://github.com/golang/go/wiki/Modules) for dependency
+management and vendoring, so we strongly recommend using Go Modules for
+dependency management in provider codebases too, which allows the `go` tool
+to automatically understand transitive dependencies and upgrade other required
+packages accordingly. Once your provider codebase is a Go Module, you can
+use the following commands to upgrade for Terraform 0.12 compatibility:
+
+```
+go get github.com/hashicorp/terraform@v0.12.0-rc1
+go mod tidy
+go mod vendor
+```
+
+-> The above uses the `v0.12.0-rc1` tag so that providers can be upgraded
+prior to the final v0.12.0 release. Once the final release is out, we will
+update this documentation to use the `v0.12.0` tag. In practice, any
+v0.12 release at least as new as RC1 is acceptable, and if newer releases
+are available we recommend using them to obtain any subsequent fixes and
+improvements to the SDK.
+
+After all of these commands are complete, you should find your version control
+detects changes to the `go.mod` and `go.sum` files as well as various files
+in the `vendor` subdirectory. While vendoring is mot mandatory for providers,
+we still recommend using it to ensure dependencies remain consistent for now,
+until the Go team has finished deploying its new solutions for module
+distribution.
+
+With the updated SDK and its dependencies installed, you should be able to run
+your provider's tests in the usual way to see how things are working. For
+simple providers, this is likely to be all of the work required! However,
+we'd still recommend reading the following sections to learn about some specific
+situations where additional changes may be helpful or required, particularly
+if you see unexpected new test failures after upgrading.
+
+Once you consider your provider ready to release with v0.12 compatibility,
+if your provider is distributed by HashiCorp (that is, available for
+installation with `terraform init`) you must be sure to be explicit about that
```

Comment:
"you must be sure to be explicit about that" this is ambiguous as to what "that" means